mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala

```diff
@@ -0,0 +1,202 @@
+import dotty.tools.dotc.core.Denotations
+import dotty.tools.dotc.core.Denotations.Denotation
+import dotty.tools.dotc.core.Denotations.MultiPreDenotation
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.NameOps._
+import dotty.tools.dotc.core.Names
+import dotty.tools.dotc.core.Names._
+import dotty.tools.dotc.core.StdNames
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types.NamedType
+import dotty.tools.dotc.interactive.Interactive
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.interactive.SourceTree
+import dotty.tools.dotc.transform.SymUtils._
+import dotty.tools.dotc.util.NoSourcePosition
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.util.SourcePosition
+import org.eclipse.lsp4j.Location
+
+class PcDefinitionProvider(
+    driver: InteractiveDriver,
+    params: OffsetParams,
+    search: SymbolSearch
+) {
+
+  def definitions(): DefinitionResult = {
+    val uri = params.uri
+    val filePath = Paths.get(uri)
+    val diagnostics = driver.run(
+      uri,
+      SourceFile.virtual(filePath.toString, params.text)
+    )
+    val unit = driver.currentCtx.run.units.head
+    val tree = unit.tpdTree
+
+    val pos = driver.sourcePosition(params)
+    val path =
+      Interactive.pathTo(driver.openedTrees(uri), pos)(using driver.currentCtx)
+    given ctx: Context = driver.localContext(params)
+    findDefinitions(tree, path, pos, driver)
+  }
+
+  private def findDefinitions(
+      tree: Tree,
+      path: List[Tree],
+      pos: SourcePosition,
+      driver: InteractiveDriver
+  )(using ctx: Context): DefinitionResult = {
+    enclosingSymbols(path, pos) match {
+      case symbols @ (sym :: other) =>
+        val semanticSym = SemanticdbSymbols.symbolName(sym)
+        val isLocal = sym.source == pos.source
+        if (isLocal)
+          val defs =
+            Interactive.findDefinitions(List(sym), driver, false, false)
+          defs.headOption match {
+            case Some(srcTree) =>
+              val pos = srcTree.namePos
+              location(pos) match {
+                case None => DefinitionResultImpl.empty
+                case Some(loc) =>
+                  DefinitionResultImpl(
+                    semanticSym,
+                    List(loc).asJava
+                  )
+              }
+            case None =>
+              DefinitionResultImpl.empty
+          }
+        else {
+          val res = new ArrayList[Location]()
+          symbols
+            .sorted(symOrdering)
+            .map(SemanticdbSymbols.symbolName)
+            .foreach { sym =>
+              res.addAll(search.definition(sym, params.uri()))
+            }
+          DefinitionResultImpl(
+            SemanticdbSymbols.symbolName(sym),
+            res
+          )
+        }
+      case Nil => DefinitionResultImpl.empty
+    }
+  }
+  assert(true)
+
+  private def location(p: SourcePosition): Option[Location] = {
```

Comment:
We already have `def location` in the ScalaPresentationCompiler, could you move it to extension and move in both places?

Reply:
moved to extension